### DISCUSSION TITLE
RT-Thread编程陷阱与防坑指南清单

### DISCUSSION THREAD

**Discussion by geniusgogo** [RT-Thread/rt-thread#6225] [Show and tell]
此issue登记记录在开发过程中遇到的一些使用陷阱，并最好能提供如何规避此类陷阱的方法、思路以及故事:) 

* rt_malloc/rt_free
> - 内核版本：<4.0.3
> - 陷阱1：不应该出现在中断上下文中
>> - 说明：由于rt_malloc/rt_free里会操作锁，RT-Thread在中断上下文中不可拿锁。
> - 陷阱2：不应该在关调度器、关中断后使用
>> - 说明：锁有可能被其它任务占用，此时关了调度器或者中断后，本任务就可能拿锁失败，锁死系统

**Comment by mysterywolf** [comment]
# 双精度浮点问题
部分STM32高级芯片具备双精度浮点计算能力，设置方式有所区别，参见：
https://club.rt-thread.org/ask/question/430612.html
https://github.com/RT-Thread/rt-thread/issues/4186 ->  https://github.com/RT-Thread/rt-thread/pull/4188
https://github.com/RT-Thread-Studio/sdk-bsp-stm32h750-realthread-artpi/issues/177 -> https://github.com/RT-Thread-Studio/sdk-bsp-stm32h750-realthread-artpi/pull/195

**Comment by mysterywolf** [comment]
# 慎用RT_IPC_FLAG_FIFO
除非你有明确理由，非常关心内核对象等待队列排列顺序的先来后到，否则强烈建议用RT_IPC_FLAG_PRIO

**Comment by haodongnj** [comment]
# STM32 CAN Bus 的比特率
STM32的CAN驱动里面代码适配时，有预设的总线时钟。需要核查驱动预设的时钟和你使用的板子的总线时钟是否匹配。如果不匹配，则需要自行修改驱动里面的时钟配置选项，否则CAN总线的bitrate会不合预期。

**Comment by enkiller** [comment]
# scons Windows 平台编译，命令行参数超过 32K 报错规避办法

1. 打开工程目录下 `SConstruct` 文件：

2. 在 `objs = PrepareBuilding(env, RTT_ROOT)` 这一行前面添加如下代码

```python
def ourspawn(sh, escape, cmd, args, e):
    newargs = ' '.join(args[1:])
    cmdline = cmd + " " + newargs
    if (len(cmdline) > 16 * 1024):
        f = open('cmd.param', 'w')
        f.write(' '.join(args[1:]).replace('\\', '/'))
        f.close()
        # exec
        cmdline = cmd + " @cmd.param"
    proc = subprocess.Popen(cmdline, stdin=subprocess.PIPE, stdout=subprocess.PIPE,
        stderr=subprocess.PIPE, shell = False, env = e)
    data, err = proc.communicate()
    rv = proc.wait()
    if rv != 0:
        sys.stderr.write(err)
    else:
        sys.stdout.write(data)
    return rv

if platform.system() == 'Windows':
    env['SPAWN'] = ourspawn
```

3. 在文件头部导入相应的模块

```python
import subprocess
import platform
```

**Comment by supperthomas** [comment]
[rt-thread 使用宝典](https://club.rt-thread.org/ask/article/2460fcd7db4821ae.html)
